src/witan/models/dem/ccm/components_functions.clj

```diff
@@ -170,6 +205,48 @@
                        (range (inc first-proj-year) (inc last-proj-year))))
               [:sex :year :age])))
 
+(defn apply-national-trend-fertility
+  "Takes a dataset of projected rates or values for the jumpoff year, and a dataset of
+   projected national rates for the following years of the projection (currently
+   this is ONS data). Takes parameters for first & last projection years, the
+   scenario to be used from the future rates dataset (e.g. :low, :principal, or
+   :high), and the name of the column with rates or value in the jumpoff year dataset
+   (e.g. :fert-rate or :births). Returns a dataset with projected rates or values
+   for each age, sex, and projection year, calculated by applying the trend from
+   the projected national rates to the data from the jumpoff year dataset."
+  [jumpoff-year-projection future-assumption
+   first-proj-year last-proj-year scenario assumption-col]
+  (let [proportional-differences (calc-proportional-differences-fertility future-assumption
+                                                                          first-proj-year
+                                                                          last-proj-year
+                                                                          scenario)
+        jumpoff-year-projection-n (wds/row-count jumpoff-year-projection)
+        projection-first-proj-year (-> jumpoff-year-projection
+                                       (ds/add-column :year (repeat jumpoff-year-projection-n
+                                                                    first-proj-year))
+                                       (ds/select-columns [:gss-code :sex :age :year assumption-col]))]
+    (order-ds (:accumulator
+               (reduce (fn [{:keys [accumulator last-calculated]} current-year]
+                         (let [projection-this-year (-> proportional-differences
+                                                        (wds/select-from-ds {:year current-year})
+                                                        (wds/join last-calculated [:age])
+                                                        (wds/add-derived-column :prop-diff
+                                                                                [:prop-diff]
+                                                                                (fn [pd] (case (nil? pd)
```

Comment:
`case` is a bit much for essentially a nil check. Also, consider "nil punning":
`#(or % 0.0)` - will return `%` if not nil, else `0.0`